Ufs hera update

AlexanderRichert-NOAA opened the pull request:
This pull request contains the changes needed to build the UFS WM on Hera, including compiling static libraries with the same versions as hpc-stack. Most changes are related to enabling static libraries throughout the whole stack.

climbfuji commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -88,6 +88,7 @@ class Esmf(MakefilePackage):
         when="@8.3.0b09",
     )
     variant("debug", default=False, description="Make a debuggable version of the library")
+    variant("caplinks", default=False, description="Create symlinks libESMF.{a,so}")
```
I don't think we should do this. We need to figure out how the `-lESMF` comes in, and fix it.

climbfuji commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -311,6 +313,14 @@ def edit(self, spec, prefix):
             os.environ["ESMF_NETCDF"] = "nc-config"
             os.environ["ESMF_NFCONFIG"] = "nf-config"
 
+            if spec["hdf5"].satisfies("~shared"):
```
This shouldn't be needed either, `ESMF-NETCDF` and `ESMF_NFCONFIG` should be sufficient?

climbfuji commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -360,6 +370,12 @@ def edit(self, spec, prefix):
     def install_findesmf(self):
         install_tree("cmake", self.prefix.cmake)
 
+    @run_after("install")
```
As above

climbfuji commented on var/spack/repos/builtin/packages/bacio/package.py:
```diff
@@ -39,3 +39,6 @@ def cmake_args(self):
         return args
+
+    def patch(self):
+        filter_file('2\.4\.0',str(self.spec.version),"VERSION",when="bacio@2.4.1")
```
Version 2.4.0 is hardcoded?

AlexanderRichert-NOAA replied:
Yes, in the "VERSION" file in bacio repo

climbfuji commented on var/spack/repos/builtin/packages/netcdf-c/package.py:
```diff
@@ -206,6 +213,11 @@ def configure_args(self):
         if hdf5_hl.satisfies("~shared"):
             libs.append(hdf5_hl.libs.link_flags)
 
+        if "~shared" in self.spec:
```
These should come in via `hdf5_hl.libs.link_flags`? Same as for `szlib`, `zstd`, ...

climbfuji commented on var/spack/repos/builtin/packages/netcdf-fortran/package.py:
```diff
@@ -33,8 +33,13 @@ class NetcdfFortran(AutotoolsPackage):
+    depends_on("netcdf-c+shared", when="+shared")
     depends_on("netcdf-c@4.7.4:", when="@4.5.3:")  # nc_def_var_szip required
     depends_on("doxygen", when="+doc", type="build")
+    depends_on("mpi", when="^netcdf-c+parallel-netcdf")
```
Lines 39-42 should not be needed. netcdf-c depends on MPI and hdf5, that should be sufficient? We've been building parallel netcdf libraries with Fortran API so far, and didn't have any problems.

climbfuji commented on var/spack/repos/builtin/packages/netcdf-fortran/package.py:
```diff
@@ -132,6 +137,32 @@ def configure_args(self):
                 # not run by default and explicitly disabled above. To avoid the
                 # configuration failure, we set the following cache variable:
                 config_args.append("ac_cv_func_MPI_File_open=yes")
+            if "~shared" in self.spec:
```
We will never be able to get lines 140 to 165 merged into the authoritative spack repository. Each of these packages should identify its own library dependencies and flags?

climbfuji commented on var/spack/repos/jcsda-emc-bundles/packages/ufs-weather-model-env/package.py:
```diff
@@ -12,32 +12,51 @@ class UfsWeatherModelEnv(BundlePackage):
-    depends_on("mapl+debug", type="run", when="+debug")
-
-    # There is no need for install() since there is no code.
+    depends_on("bacio@2.4.1")
```
We've tried to not encode versions in these bundle packages and leave that up to the common packages.yaml and the templates. I am not sure if changing this for the ufs-weather-model-env only but keeping the rest as-is is what we should do. Is there a reason why the versions/variants can't be in configs/templates/ufs-weather-model/spack.yaml?

I am open to revisit how/where we specify versions and variants, but we should be consistent across the various templates, bundle packages, etc.

climbfuji commented on var/spack/repos/jcsda-emc/packages/mapl/package.py:
```diff
@@ -62,12 +62,15 @@ class Mapl(CMakePackage):
     variant("debug", default=False, description="Make a debuggable version of the library")
     variant("extdata2g", default=False, description="Use ExtData2G")
+    variant("pnetcdf", default=True, description="Use parallel netCDF")
+    variant("buildexe", default=True, description="Build pfio test executables")
```
Any reason why you may want to turn this off?

climbfuji reviewed (changes requested):
I am concerned about some of the changes in this PR. They look like a lot of duct-tape solutions for underlying problems that should be fixed in other places. I am worried that we'll not be able to get this merged into the authoritative/upstream repositories, which is I think what we want. For the bundle packages (which we will never send upstream), we'll need to come up with a consistent solution, I am open for discussion.

But, in general, what forces us to use static libraries throughout? Can't we use dynamic netcdf libraries. Even stone-age cray systems nowadays support dynamic libraries just fine.

AlexanderRichert-NOAA commented:
> But, in general, what forces us to use static libraries throughout? Can't we use dynamic netcdf libraries. Even stone-age cray systems nowadays support dynamic libraries just fine.

The motivation for static libraries is mainly consistency with hpc-stack. Also, as far as I know, static libraries are still required for operations on WCOSS2.

AlexanderRichert-NOAA commented:
I've confirmed with EMC that static libraries are required. I don't think it should be too difficult to make this happen, but it'll take some more elbow grease to figure out a smoother way to make some of the changes (namely linker flag-related package tweaks). Depending on what's involved in resolving the -lESMF and MAPL/pfio exe issues (which I think are my lingering two "hack-y" fixes outside of linker flags/enabling static builds), I may push for quick-and-dirty fixes in the short term to let UFS developers at least get running with spack stack, with the expectation that we revert those changes once permanent solutions are found.

climbfuji commented:
> I've confirmed with EMC that static libraries are required. I don't think it should be too difficult to make this happen, but it'll take some more elbow grease to figure out a smoother way to make some of the changes (namely linker flag-related package tweaks). Depending on what's involved in resolving the -lESMF and MAPL/pfio exe issues (which I think are my lingering two "hack-y" fixes outside of linker flags/enabling static builds), I may push for quick-and-dirty fixes in the short term to let UFS developers at least get running with spack stack, with the expectation that we revert those changes once permanent solutions are found.

Thanks, yes, that sounds good.

AlexanderRichert-NOAA commented:
Update:

- For now I have removed the -lESMF-related tweak, as this can be fixed through UFS. If needed, we have a couple ways to fix it within spack-stack (symlinks, or disable installation of ESMF's cmake directory).
- I have added a patch to fix HDF5's pkgconfig issues (which it turns out are numerous). An issue has been created for HDF5.
- The remaining "ugly" modification is for MAPL. Under hpc-stack, MAPL uses esma_cmake's FindBaselibs.cmake, which calls nf-config and hands off the necessary flags to compile for the static case. Unfortunately, MAPL does not appear to provide a nice way to specify netcdf-related compile flags. The current solution adds the appropriate output of nc-config to pfio/CMakeLists.txt. Alternative solutions could include adding esma_cmake to spack-stack, or creating a similar cmake file providing FindBaselibs.cmake's functionality as a patch.

climbfuji commented:
> Update:
> 
> * For now I have removed the -lESMF-related tweak, as this can be fixed through UFS. If needed, we have a couple ways to fix it within spack-stack (symlinks, or disable installation of ESMF's cmake directory).
> * I have added a patch to fix HDF5's pkgconfig issues (which it turns out are numerous). An issue has been created for HDF5.
> * The remaining "ugly" modification is for MAPL. Under hpc-stack, MAPL uses esma_cmake's FindBaselibs.cmake, which calls nf-config and hands off the necessary flags to compile for the static case. Unfortunately, MAPL does not appear to provide a nice way to specify netcdf-related compile flags. The current solution adds the appropriate output of nc-config to pfio/CMakeLists.txt. Alternative solutions could include adding esma_cmake to spack-stack, or creating a similar cmake file providing FindBaselibs.cmake's functionality as a patch.

@mathomp4 What do you think, how should we/the ufs use MAPL?

climbfuji commented on var/spack/repos/jcsda-emc-bundles/packages/ufs-weather-model-env/package.py:
```diff
@@ -22,22 +22,39 @@ class UfsWeatherModelEnv(BundlePackage):
-
-    # There is no need for install() since there is no code.
+    depends_on("bacio")
+    depends_on("fms")
```
Removing the `fms` and `crtm` versions here will not work. If you install the jedi-ufs-all environment, spack may decide to use the JEDI versions of both libraries (`~release-jcsda`). The only way to install both `ufs-weather-model-env` and `jedi-fv3-env` in the same environment and get the correct versions of `fms` and `crtm` loaded for each of them is to hardcode them here.

climbfuji commented on var/spack/repos/jcsda-emc-bundles/packages/ufs-weather-model-env/package.py:
```diff
@@ -22,22 +22,39 @@ class UfsWeatherModelEnv(BundlePackage):
+    depends_on("netcdf-fortran+shared", when="+shared")
+    depends_on("zlib+shared", when="+shared")
+
+    depends_on(f"esmf~debug", when="~debug")
```
Why are you using `f` strings in lines 55-60, but not above?

climbfuji commented on var/spack/repos/jcsda-emc-bundles/packages/ufs-weather-model-env/package.py:
```diff
@@ -22,22 +22,39 @@ class UfsWeatherModelEnv(BundlePackage):
+    depends_on("ip")
+    depends_on("sp")
+    depends_on("w3emc")
+    depends_on("parallelio+fortran~pnetcdf~shared")
```
Wondering if this can be simplified ... create a variant `shared` for the `ufs-weather-model-env` which can be `False` by default. Then:
```
if "~shared" in self.spec:
   variant = "~shared"
else:
   variant = "+shared"
depends_on("parallelio+fortran~pnetcdf{}".format(variant))
```
I haven't actually tried this, but would be interesting to see if it works.

mathomp4 commented:
> @mathomp4 What do you think, how should we/the ufs use MAPL?

@climbfuji In my spack testing (which I hope to get back to soon), I never had to do anything fancy for MAPL as you can see here:

https://github.com/NOAA-EMC/spack/pull/174/files#diff-6f7a9850240769a509f654ca949f65aea45cd45b3f8903664d9f891de6cc031f

I mean, I added new versions, fixed up the variants and added better versioning for the GFE packages, but nothing beyond that.

I never had to change anything for the netCDF usage. I just took whatever you all had there. Now, I had to fix up MAPL a bit, so it's possible an older version of MAPL still needs all that?

mathomp4 commented:
Ahh. Is it because they are using static libraries? I guess I'm just using the default shared library build from spack-stack and never thought about going back to static (since at some point we might want to use, say, other compressions from netCDF/HDF5 which require shared libraries).

Note that you do pull in esma_cmake in the package here but of course we "avoid" the Baselibs bits of it since you don't use baselibs with spack (obviously). But, as long as FindNetCDF can handle a static build of netCDF, I see no reason MAPL couldn't. It's just doing that:
```cmake
if (NOT Baselibs_FOUND)
  set(MPI_DETERMINE_LIBRARY_VERSION TRUE)
  find_package(MPI)


  find_package(NetCDF REQUIRED C Fortran)
```

mathomp4 commented:
And one more thing as I look over the changes, I have *never* used MAPL with pnetcdf, so that is mystery land. We've never even tried it, just done parallel netCDF via HDF5.

AlexanderRichert-NOAA commented:
> Ahh. Is it because they are using static libraries? I guess I'm just using the default shared library build from spack-stack and never thought about going back to static (since at some point we might want to use, say, other compressions from netCDF/HDF5 which require shared libraries).
> 
> Note that you do pull in esma_cmake in the package here but of course we "avoid" the Baselibs bits of it since you don't use baselibs with spack (obviously). But, as long as FindNetCDF can handle a static build of netCDF, I see no reason MAPL couldn't. It's just doing that:
> 
> ```cmake
> if (NOT Baselibs_FOUND)
>   set(MPI_DETERMINE_LIBRARY_VERSION TRUE)
>   find_package(MPI)
> 
> 
>   find_package(NetCDF REQUIRED C Fortran)
> ```

Yes, it is because of the static libraries. With everything static, the issue is that the link flags from the upstream deps (zlib, hdf5) don't get propagated into pfio/CMakeLists.txt. Is there a way to get those in there without esma_cmake?

AlexanderRichert-NOAA commented:
@edwardhartnett, the changes made in this PR work for the latest versions of all packages, with I think the exception of jasper which I left at 2.0.32. It can build all static libraries with netdf-c@4.9.0 with zstd, and UFS builds and runs. The only modification needed to UFS was to accommodate ESMF 8.4.0's use of external parallelio in the link step, but otherwise no issues that I can see based on running UFS's main control case.

For posterity, here's the spec that I mainly tested with:
`ufs-weather-model-env~debug~python~shared ^bacio@2.4.1 ^crtm@2.4.0 ^esmf@8.3.0b09~shared ^fms@2022.01 ^g2@3.4.5 ^g2tmpl@1.10.0 ^hdf5@1.10.6+hl~shared ^ip@3.3.3 ^jasper@2.0.25 ^libjpeg-turbo@2.1.0 ^libpng@1.6.37 ^mapl@2.22.0~pnetcdf~shared ^netcdf-c@4.7.4~dap~shared ^netcdf-fortran@4.5.4~shared ^parallelio@2.5.7~shared ^sp@2.3.3 ^w3emc@2.9.2 ^libxml2@2.9.13 ^zlib@1.2.11`

And the more limited testing I did with recent package versions:
`ufs-weather-model-env~debug~python~shared ^bacio@2.5.0 ^crtm@2.4.0 ^esmf@8.4.0~shared+parallelio+mpi ^fms@2022.03 ^g2@3.4.5 ^g2tmpl@1.10.2 ^hdf5@1.13.2+hl~shared ^ip@4.0.0 ^jasper@2.0.32 ^libjpeg-turbo@2.1.3 ^libpng@1.6.37 ^mapl@2.22.0~pnetcdf~shared ^netcdf-c@4.9.0~dap~shared ^netcdf-fortran@4.6.0~shared ^parallelio@2.5.9~shared ^sp@2.3.3 ^w3emc@2.9.2 ^libxml2@2.9.13 ^zlib@1.2.11 ^zstd@1.5.2 libs=static`

climbfuji commented:
@AlexanderRichert-NOAA Looks like this PR is ready, before we can merge we'll need a spack-stack PR to run the CI tests with this branch.

mathomp4 commented:
> Yes, it is because of the static libraries. With everything static, the issue is that the link flags from the upstream deps (zlib, hdf5) don't get propagated into pfio/CMakeLists.txt. Is there a way to get those in there without esma_cmake?

@AlexanderRichert-NOAA New question. Which `FindNetCDF.cmake` are you using? I do see that the [one from EMC](https://github.com/NOAA-EMC/CMakeModules/blob/cabd7753ae17f7bfcc6dad56daf10868aa51c3f4/Modules/FindNetCDF.cmake) has some updates via https://github.com/NOAA-EMC/CMakeModules/pull/60 that I don't have in the `FindNetCDF.cmake` in ESMA_cmake. And those differences are all due to NetCDF-C in static mode where it calls `nc-config --libs --static` just as you did.

If you are picking up the ESMA_cmake file, could you try swapping out the EMC one? If that works, then I could patch up ESMA_cmake to use that.

climbfuji commented:
Yes, I remember we had to put those bug fixes into NOAA-EMC / CMakeModules to correctly find static netcdf libraries

> On Nov 22, 2022, at 10:58 AM, Matthew Thompson ***@***.***> wrote:
> 
> 
> Yes, it is because of the static libraries. With everything static, the issue is that the link flags from the upstream deps (zlib, hdf5) don't get propagated into pfio/CMakeLists.txt. Is there a way to get those in there without esma_cmake?
> 
> @AlexanderRichert-NOAA <https://github.com/AlexanderRichert-NOAA> New question. Which FindNetCDF.cmake are you using? I do see that the one from EMC <https://github.com/NOAA-EMC/CMakeModules/blob/cabd7753ae17f7bfcc6dad56daf10868aa51c3f4/Modules/FindNetCDF.cmake> has some updates via NOAA-EMC/CMakeModules#60 <https://github.com/NOAA-EMC/CMakeModules/pull/60> that I don't have in the FindNetCDF.cmake in ESMA_cmake. And those differences are all due to NetCDF-C in static mode where it calls nc-config --libs --static just as you did.
> 
> If you are picking up the ESMA_cmake file, could you try swapping out the EMC one? If that works, then I could patch up ESMA_cmake to use that.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/NOAA-EMC/spack/pull/192#issuecomment-1324051204>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB5C2ROCU6ZOUQWV36M6LWLWJUCVPANCNFSM6AAAAAAR2X45LY>.
> You are receiving this because you were mentioned.
>

mathomp4 commented:
Oh wait. Our FindNetCDF is from our ecbuild fork, but Spack doesn't know about our ecbuild, only esma_cmake. And my guess is the ecbuild version is probably behind the @climbfuji version EMC has.

AlexanderRichert-NOAA commented:
> Oh wait. Our FindNetCDF is from our ecbuild fork, but Spack doesn't know about our ecbuild, only esma_cmake. And my guess is the ecbuild version is probably behind the @climbfuji version EMC has.

I didn't notice until just now that the mapl spack package pulls in ESMA_cmake as a resource, but I don't see it getting used. Do you know how that's getting used? I tried adding the directory to CMAKE_PREFIX_PATH but no luck.

mathomp4 commented:
> > Oh wait. Our FindNetCDF is from our ecbuild fork, but Spack doesn't know about our ecbuild, only esma_cmake. And my guess is the ecbuild version is probably behind the @climbfuji version EMC has.
> 
> I didn't notice until just now that the mapl spack package pulls in ESMA_cmake as a resource, but I don't see it getting used. Do you know how that's getting used? I tried adding the directory to CMAKE_PREFIX_PATH but no luck.

@AlexanderRichert-NOAA This code in MAPL:
```cmake
# mepo can now clone subrepos in three styles
set (ESMA_CMAKE_DIRS
  ESMA_cmake
  @ESMA_cmake
  ESMA_cmake@
  )

foreach (dir IN LISTS ESMA_CMAKE_DIRS)
  if (EXISTS ${CMAKE_CURRENT_LIST_DIR}/${dir})
    list (APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/${dir}")
    set (ESMA_CMAKE_PATH "${CMAKE_CURRENT_LIST_DIR}/${dir}" CACHE PATH "Path to ESMA_cmake code")
    include (esma)
    set(MAPL_STANDALONE TRUE)
  endif ()
endforeach ()
```
Adds `ESMA_cmake` to the `CMAKE_MODULE_PATH` if it's found (the variant names are due to various ways mepo can clone). 

`ESMA_cmake` is brought in by:

```python
    # Versions later than 3.14 remove FindESMF.cmake
    # from ESMA_CMake. This works with mapl@2.22.0:
    resource(
        name="esma_cmake",
        git="https://github.com/GEOS-ESM/ESMA_cmake.git",
        tag="v3.17.0",
        when="@2.22.0:",
    )
```

And you'll see spack will just put it in the MAPL source as it builds:

```
	destination: /discover/nobackup/mathomp4/spack-stack/cache/build_stage/spack-stage-mapl-2.30.3-tr2j3bqel7lk6pxvdxcqshqdzhotvyqp/spack-src/ESMA_cmake
```

AlexanderRichert-NOAA commented:
> Oh wait. Our FindNetCDF is from our ecbuild fork, but Spack doesn't know about our ecbuild, only esma_cmake. And my guess is the ecbuild version is probably behind the @climbfuji version EMC has.

Yep, the difference appears to be the use of the EMC version of FindNetCDF.cmake. Is there a straightforward way to incorporate that short of having to make EMC's cmakemodules its own package?

climbfuji commented:
Note that jedi-cmake, which is already a package in spack, uses the NOAA-EMC CMakeModules (as a submodule).

AlexanderRichert-NOAA commented:
> Note that jedi-cmake, which is already a package in spack, uses the NOAA-EMC CMakeModules (as a submodule).

Yeah that almost works, the problem is that I can't get the jedi-cmake FindNetCDF.cmake to take precedence over the ecbuild one. I believe this is because ecbuild automatically modifies CMAKE_MODULE_PATH. I think I'll keep my current solution, as I don't think any other short-term approach (like patching ecbuild) will be any less ugly or risky. I think the long-term solution will either be to modify MAPL in its use of ecbuild, or to update ecbuild to support static NetCDF.

mathomp4 commented:
> > Note that jedi-cmake, which is already a package in spack, uses the NOAA-EMC CMakeModules (as a submodule).
> 
> Yeah that almost works, the problem is that I can't get the jedi-cmake FindNetCDF.cmake to take precedence over the ecbuild one. I believe this is because ecbuild automatically modifies CMAKE_MODULE_PATH. I think I'll keep my current solution, as I don't think any other short-term approach (like patching ecbuild) will be any less ugly or risky. I think the long-term solution will either be to modify MAPL in its use of ecbuild, or to update ecbuild to support static NetCDF.

@AlexanderRichert-NOAA I'll look at at least updating the FindNetCDF.cmake in the GEOS-ESM fork of ecbuild to be the same as JCSDA's...but this won't help in the end.

The unfortunate thing is that the main ecbuild `FindNetCDF.cmake` and the JCSDA one seem to be quite different. Maybe they have equivalent functionality in the end, but it's not a quick and easy comparison.

Is there a way in UFS to inject jedi-cmake's `CMAKE_MODULE_PATH` to be higher in precedence than ecbuild? Even if it is just heavy-handed?

AlexanderRichert-NOAA commented:
> Is there a way in UFS to inject jedi-cmake's `CMAKE_MODULE_PATH` to be higher in precedence than ecbuild? Even if it is just heavy-handed?

I've been unable to find a way to. I think it has to do with the fact that ecbuild itself is modifying CMAKE_MODULE_PATH, so I think it would involve modifying the ecbuild cmake code to prepend vs. append paths or vice versa...

climbfuji commented:
@AlexanderRichert-NOAA Thanks for all these efforts. I think we'll leave it as-is and create the uppercase symlinks to get past this problem for now. We'll need an issue to remind us of the hack so that we can revisit it at a later stage. What is the status of this PR otherwise? Is it ready? Asking because I have many changes in the pipeline, some in form of PRs, some as branches only so far, and all these need to go in soon (like yours I guess).

AlexanderRichert-NOAA commented:
> @AlexanderRichert-NOAA Thanks for all these efforts. I think we'll leave it as-is and create the uppercase symlinks to get past this problem for now. We'll need an issue to remind us of the hack so that we can revisit it at a later stage. What is the status of this PR otherwise? Is it ready? Asking because I have many changes in the pipeline, some in form of PRs, some as branches only so far, and all these need to go in soon (like yours I guess).

This is a different ESMF issue, which is solved by the patch I've put in mapl/package.py. For the symlinks I was planning to wait to see what EMC/UFS devs say. I just noticed a couple small items I need to fix in the next few minutes, then this PR is good to go.

climbfuji commented on var/spack/repos/jcsda-emc-bundles/packages/base-env/package.py:
```diff
@@ -18,6 +18,9 @@ class BaseEnv(BundlePackage):
 
     version("1.0.0")
 
+    variant("python", default=True, description="Include Python")
```
The Python variant doesn't make sense to me. There is a `depends_on(Python)` further down, which currently is true wether this variant is set or not. I'd remove line 21.

climbfuji commented on var/spack/repos/jcsda-emc-bundles/packages/ufs-weather-model-env/package.py:
```diff
@@ -22,8 +22,13 @@ class UfsWeatherModelEnv(BundlePackage):
     variant("python", default=True, description="Build Python dependencies")
+    variant("shared", default=True, description="Build dynamic libraries when possible")
+
+    depends_on("base-env+shared+python", type="run", when="+shared+python")
```
Likewise, get rid of all the `+python` and `~python` here.

climbfuji commented on var/spack/repos/jcsda-emc-bundles/packages/base-env/package.py:
```diff
@@ -27,12 +30,17 @@ class BaseEnv(BundlePackage):
+    depends_on("netcdf-fortran~shared", type="run", when="~shared")
+    depends_on("parallel-netcdf+shared", type="run", when="+shared")
+    depends_on("parallel-netcdf~shared", type="run", when="~shared")
+    depends_on("parallelio+fortran~pnetcdf~shared", type="run")
```
This line needs a switch for `+shared`. We should fix the logic in the ufs-weather-model build and remove the explicit request for `STATIC` there, just take whatever spack-stack has built.

climbfuji reviewed (changes requested):
I think that the entire, messy +shared ~shared logic can be simplified, but we can do this in a follow-up PR. There are a few things that need to be addressed before you update the submodule pointer in your spack-stack PR and we rerun the tests.

AlexanderRichert-NOAA commented:
@climbfuji Let me know what you think re: parallelio and leaving it static-only for now to not break existing functionality (just created https://github.com/ufs-community/ufs-weather-model/issues/1512). In any case I'll need help setting the spack-stack module. The goal is to not change any existing behavior (in terms of existing default "shared" libs etc.), so fingers crossed the CI tests shouldn't be affected...

climbfuji commented on var/spack/repos/builtin/packages/netcdf-c/package.py:
```diff
@@ -153,31 +161,31 @@ def configure_args(self):
+        config_args.extend(self.enable_or_disable("parallel-tests"))
+        config_args.extend(self.enable_or_disable("cdf5"))
+        config_args.extend(self.enable_or_disable("netcdf-4"))
+        config_args.extend(self.enable_or_disable("doxygen"))
```
This does not work on my mac. Even though the default for `doxygen` is false and I don't see any place where it requested via `+doxygen`, spack's solver comes out with `+doxygen` in the spec, the configure line has `--enable-doxygen` and the error is:
```
Making all in docs
Making all in images
make[3]: Nothing to be done for `all'.
Doxyfile \
	cp auth.md obsolete/fan_utils.html html
make[3]: Doxyfile: No such file or directory
make[3]: *** [doxyfile.stamp] Error 1
make[2]: *** [all-recursive] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
==> Error: ProcessError: Command exited with status 2:
    'make' '-j10' 'V=1'
```
Why do we need to build netcdf-c documentation at all? Clearly it's available on the web and we are not trying to develop netcdf-c within spack. Can we just remove the doxygen lines 176 and 86 altogether?
```

climbfuji commented on var/spack/repos/builtin/packages/nccmp/package.py:
```diff
@@ -28,4 +30,18 @@ def cmake_args(self):
+                args.append(
+                  self.define(
+                    "CMAKE_C_COMPILER",
+                    self.spec["mpi"].mpicc
```
I am getting this error on Cheyenne:
```
/glade/work/heinzell/spack-stack-tst-120/spack-stack/spack/var/spack/repos/builtin/packages/nccmp/package.py:40, in cmake_args:
         37            args.append(self.define("CMAKE_EXE_LINKER_FLAGS", nc_flags))
         38
         39        if self.spec.satisfies("^mpi"):
  >>     40            args.append(self.define("CMAKE_C_COMPILER", self.spec["mpi"].mpicc))
         41
         42        return args
```
Again a weird one since I don't see `~shared` in the netcdf-c flags.

climbfuji reviewed (commented):
I am getting failures with this PR on my mac and on Cheyenne.

climbfuji commented:
I saw you made updates after I started to test this PR, therefore the errors I've been seeing didn't make sense. Let me try to pull in your latest two bug fix commits and see what happens.

mathomp4 commented on var/spack/repos/builtin/packages/netcdf-c/package.py:
```diff
@@ -161,31 +153,31 @@ def configure_args(self):
+        config_args = [
+            "--enable-v2",
+            "--enable-utilities",
+            "--enable-static",
```
Wouldn't this hardcode in static libraries? Why can't the choice of static libraries be an option to the package?

mathomp4 reviewed (approved):
I think this is good, though I am concerned we might have a lot of code in `builtin` that the mother spack repo will never take.

climbfuji commented:
> I think this is good, though I am concerned we might have a lot of code in `builtin` that the mother spack repo will never take.

That is my concern, too. But the burden will be on @AlexanderRichert-NOAA (sorry, Alex) to propose those changes to the authoritative repo. It is quite likely that they don't reject the changes altogether, but that they come up with alternative solutions, which we then have to bring back to our fork. But in general this should help to improve our build system.
